app/src/tracing.h

```diff
@@ -0,0 +1,222 @@
+      client_request_id.value_or(""), \
+      request_id, \
+      ##__VA_ARGS__); \
+  else \
+    CCF_APP_FAIL("RequestId={} " s, request_id, ##__VA_ARGS__)
+
+  template <typename Fn, typename Ctx>
+  Fn generic_tracing_adapter(
+    Fn fn, const std::function<ccf::ApiResult(::timespec& time)>& get_time)
+  {
+    return [fn, get_time](Ctx& ctx) {
+      auto cleanup = finally(clear_trace_state);
+
+      request_id = create_request_id();
+      ctx.rpc_ctx->set_response_header(REQUEST_ID_HEADER, request_id);
+
+      client_request_id =
+        ctx.rpc_ctx->get_request_header(CLIENT_REQUEST_ID_HEADER);
+
+      if (client_request_id.has_value())
+      {
+        // Validate client request id to avoid misinterpretation of the log,
+        // e.g. if it contains a space.
+        std::smatch match;
+        if (!std::regex_match(
+              client_request_id.value(), match, CLIENT_REQUEST_ID_REGEX))
+        {
+          client_request_id = std::nullopt;
+          SCITT_INFO("Code=InvalidInput Invalid client request id");
+          ctx.rpc_ctx->set_error(
+            HTTP_STATUS_BAD_REQUEST,
+            errors::InvalidInput,
+            "Invalid client request id.");
+          return;
+        }
+        ctx.rpc_ctx->set_response_header(
+          "x-ms-client-request-id", client_request_id.value());
+      }
+
+      ctx.rpc_ctx->set_user_data(
+        std::make_shared<AppData>(AppData{request_id, client_request_id}));
+
+      auto query = ctx.rpc_ctx->get_request_query();
+
+      if (query.empty())
+      {
+        SCITT_INFO(
+          "Verb={} Path={}",
+          ctx.rpc_ctx->get_request_verb().c_str(),
+          ctx.rpc_ctx->get_request_path());
+      }
+      else
+      {
+        SCITT_INFO(
+          "Verb={} Path={} Query={}",
+          ctx.rpc_ctx->get_request_verb().c_str(),
+          ctx.rpc_ctx->get_request_path(),
+          ctx.rpc_ctx->get_request_query());
+      }
+
+      ::timespec start;
+      ccf::ApiResult result = get_time(start);
+      if (result != ccf::ApiResult::OK)
+      {
+        SCITT_FAIL("Code=InternalError get_untrusted_host_time_v1 failed");
+        ctx.rpc_ctx->set_error(
+          HTTP_STATUS_INTERNAL_SERVER_ERROR,
+          errors::InternalError,
+          "Failed to get time.");
+        return;
+      }
+
+      fn(ctx);
+
+      ::timespec end;
+      result = get_time(end);
+      if (result != ccf::ApiResult::OK)
+      {
+        SCITT_FAIL("Code=InternalError get_untrusted_host_time_v1 failed");
+        ctx.rpc_ctx->set_error(
+          HTTP_STATUS_INTERNAL_SERVER_ERROR,
+          errors::InternalError,
+          "Failed to get time.");
+        return;
+      }
+
+      auto duration_ms = diff_timespec_ms(start, end);
+
+      if (query.empty())
+      {
+        SCITT_INFO(
+          "Verb={} Path={} Status={} TimeMs={}",
+          ctx.rpc_ctx->get_request_verb().c_str(),
+          ctx.rpc_ctx->get_request_path(),
```

Comment:
If a request includes path parameters, they are going to be included in this. This might make aggregation and analysis of logs difficult since we won't be able to group directly by path.

Maybe we should put the "method name", eg. `/entries/{txid}`, and then include the parameters as a separate field. (The CCF logging example does something like that, logging the method rather than the path)

Reply:
Good point, ok, I think I'll add `make_endpoint` after all in this PR.